src/main/kotlin/io/zerobase/smarttracing/resources/PdfGenerator.kt

```diff
@@ -0,0 +1,131 @@
+        } catch (e: IOException) {
+            println("Could not generate QR Code, IOException :: " + e.message)
+        }
+        // We set-up a Thymeleaf rendering engine. All Thymeleaf templates
+// are HTML-based files located under "src/test/resources". Beside
+// of the main HTML file, we also have partials like a footer or
+// a header. We can re-use those partials in different documents.
+        val templateResolver = ClassLoaderTemplateResolver()
+        templateResolver.prefix = "/"
+        templateResolver.suffix = ".html"
+        templateResolver.templateMode = TemplateMode.HTML
+        templateResolver.characterEncoding = UTF_8
+        val templateEngine = TemplateEngine()
+        templateEngine.setTemplateResolver(templateResolver)
+        val data = organizationData(businessname, town, state)
+        val context = Context()
+        val dataMap = HashMap<String, Any>()
+        dataMap["data"] = data
+        dataMap["logoPath"] = QR_CODE_IMAGE_PATH
+        context.setVariables(dataMap)
+        // Flying Saucer needs XHTML - not just normal HTML. To make our life
+// easy, we use JTidy to convert the rendered Thymeleaf template to
+// XHTML. Note that this might not work for very complicated HTML. But
+// it's good enough for a simple letter.
+
+        val renderedHtmlContent = templateEngine.process("template", context)
+        val xHtml = convertToXhtml(renderedHtmlContent)
+        val renderer = ITextRenderer()
+        // FlyingSaucer has a working directory. If you run this test, the working directory
+// will be the root folder of your project. However, all files (HTML, CSS, etc.) are
+// located under "/src/test/resources". So we want to use this folder as the working
+// directory.
+        val baseUrl = FileSystems
+                .getDefault()
+                .getPath("src", "main", "resources", "pdfconfig")
+                .toUri()
+                .toURL()
+                .toString()
+        print(baseUrl);
+        renderer.setDocumentFromString(xHtml, baseUrl)
+        renderer.layout()
+        // And finally, we create the PDF:
+        val outputStream: OutputStream = FileOutputStream(OUTPUT_FILE)
+        renderer.createPDF(outputStream)
+        outputStream.close()
+    }
+
+    @Throws(UnsupportedEncodingException::class)
+    private fun convertToXhtml(html: String): String {
+        val tidy = Tidy()
+        tidy.inputEncoding = UTF_8
+        tidy.outputEncoding = UTF_8
+        tidy.xhtml = true
+        val inputStream = ByteArrayInputStream(html.toByteArray(charset(UTF_8)))
+        val outputStream = ByteArrayOutputStream()
+        tidy.parseDOM(inputStream, outputStream)
+        return outputStream.toString(UTF_8)
+    }
+
+    private fun organizationData(businessname: String, state: String, town: String): Data {
+        val data = Data()
+        data.businessname = businessname
+        data.state = state
+        data.town = town
```

Comment:
we do not have state or town in our database, so where are we getting this from? this is also not sent in with any of the api parameters that should create the pdf (create scannable api call)

Reply:
think this was addressed in another comment by David -- keeping to the template for now but this is a good point. I think this will change as we go. Might be useful to have for now for our first planned deploy